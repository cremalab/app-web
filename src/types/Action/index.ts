import { Action as ReduxAction } from "redux"

/**
 * `Action`
 * ---------------------------------------------
 * Extends redux `Action` adding `payload`
 */

export interface Action<Type extends PropertyKey, Payload>
  extends ReduxAction<Type> {
  payload: Payload
}

/**
 * `Action`
 * ------------------
 * Returns a `Tagged<Tag, Data>`
 **/

export const Action = <Type extends PropertyKey, Payload>(
  type: Type,
  payload: Payload,
): Action<Type, Payload> => ({
  type,
  payload,
})
