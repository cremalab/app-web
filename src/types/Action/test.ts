import { Action } from "."

describe("Action", () => {
  it("returns an `Action<Type, Payload>`", () => {
    // Arrange
    const type = "SOME_TYPE"
    const payload = 2

    // Act
    const received = Action(type, payload)

    // Assert
    expect(received).toMatchSnapshot()
  })
})
