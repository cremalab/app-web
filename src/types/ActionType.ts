export enum ActionType {
  TodoAdd = "TodoAdd",
  TodoRemove = "TodoRemove",
  TodoUpdate = "TodoUpdate",
}
