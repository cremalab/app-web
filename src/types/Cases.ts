import { Tagged } from "./Tagged"

export type Cases<
  A extends Tagged<Tag, unknown>,
  Return,
  Tag extends PropertyKey = PropertyKey
> = {
  [K in A["tag"]]: A extends Tagged<K, infer Data> ? (x: Data) => Return : never
}
