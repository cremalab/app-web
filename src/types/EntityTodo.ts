import { Entity } from "./Entity"
import { Todo } from "./Todo"

/**
 * EntityTodo - Entity & Todo
 * ------------------------------------------------
 * A Todo with an ID
 */

export interface EntityTodo extends Entity, Todo {}
