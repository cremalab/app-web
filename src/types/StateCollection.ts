import { Entity } from "./Entity"

export interface StateCollection<A extends Entity> {
  entities: { [key: string]: A }
  result: readonly Entity["id"][]
}
