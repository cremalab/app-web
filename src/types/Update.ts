import { Entity } from "./Entity"
import { Omit } from "./Omit"

export interface Update<A extends Entity> {
  id: A["id"]
  properties: Partial<Omit<A, "id">>
}
