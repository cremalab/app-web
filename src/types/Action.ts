import { Tagged } from "./Tagged"
import { ActionType } from "./ActionType"

export interface Action<Payload = null, Type extends PropertyKey = ActionType>
  extends Tagged<Type> {
  payload: Payload
}
