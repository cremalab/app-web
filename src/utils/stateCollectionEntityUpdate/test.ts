import { stateCollectionEntityUpdate } from "."
import { EntityTodo } from "../../types/EntityTodo"
import { StateCollection } from "../../types/StateCollection"
import { Action } from "../../types/Action"
import { UpdateEntity } from "../../types/Update"

describe("stateCollectionEntityUpdate", () => {
  it("updates `Entity<A>` in `StateCollection<Entity<A>>`", () => {
    // Arrange
    const state: StateCollection<EntityTodo> = {
      entities: {
        "1": { id: "1", title: "Test Todo", complete: false },
        "2": { id: "2", title: "Test Todo", complete: false },
      },
      result: ["1", "2"],
    }

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const action: Action<any, UpdateEntity<EntityTodo>> = {
      type: "ANY",
      payload: { id: "1", properties: { complete: true } },
    }

    // Act
    const received = stateCollectionEntityUpdate(state)(action)

    // Assert
    expect(received).toMatchSnapshot()
  })
})
