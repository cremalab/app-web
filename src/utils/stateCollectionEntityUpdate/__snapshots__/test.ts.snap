// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stateCollectionEntityUpdate updates \`Entity<A>\` in \`StateCollection<Entity<A>>\` 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "complete": true,
      "id": "1",
      "title": "Test Todo",
    },
    "2": Object {
      "complete": false,
      "id": "2",
      "title": "Test Todo",
    },
  },
  "result": Array [
    "1",
    "2",
  ],
}
`;
